gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/auth/RefreshTokenAuth2Provider.java

```diff
@@ -0,0 +1,187 @@
+        TOKEN_SERVER_URL_SUFFIX.withPrefixes(CONFIG_KEY_PREFIXES).get(config, config::get);
+    RedactedString refreshToken =
+        AUTH_REFRESH_TOKEN_SUFFIX.withPrefixes(CONFIG_KEY_PREFIXES).getPassword(config);
+    String clientId =
+        AUTH_CLIENT_ID_SUFFIX.withPrefixes(CONFIG_KEY_PREFIXES).get(config, config::get);
+    RedactedString clientSecret =
+        AUTH_CLIENT_SECRET_SUFFIX.withPrefixes(CONFIG_KEY_PREFIXES).getPassword(config);
+
+    checkNotNull(refreshToken, "Must provide a refresh token");
+    checkNotNull(clientSecret, "Must provide a client secret");
+
+    logger.atFine().log(
+        "Refresh token calling endpoint '%s' with client id '%s'", tokenServerUrl, clientId);
+
+    try {
+      HttpTransport httpTransport = getTransport();
+      accessToken =
+          Optional.of(
+              getAccessToken(
+                  tokenServerUrl,
+                  clientId,
+                  clientSecret,
+                  previousRefreshToken.orElse(refreshToken),
+                  httpTransport));
+
+      logger.atFine().log(
+          "New access token expires at '%s'",
+          lazy(() -> Instant.ofEpochMilli(accessToken.get().getExpirationTimeMilliSeconds())));
+
+    } catch (IOException e) {
+      logger.atSevere().withCause(e).log("Couldn't refresh token");
+    }
+  }
+
+  private AccessTokenProvider.AccessToken getAccessToken(
+      String tokenServerUrl,
+      String clientId,
+      RedactedString clientSecret,
+      RedactedString refreshToken,
+      HttpTransport httpTransport)
+      throws IOException {
+
+    logger.atFine().log("Get a new access token using the refresh token grant flow");
+
+    TokenRequest request =
+        new RefreshTokenRequest(
+                httpTransport, JSON_FACTORY, new GenericUrl(tokenServerUrl), refreshToken.value())
+            .setClientAuthentication(
+                new ClientParametersAuthentication(clientId, clientSecret.value()));
+
+    TokenResponse tokenResponse = request.execute();
+    previousRefreshToken =
+        Optional.ofNullable(RedactedString.create(tokenResponse.getRefreshToken()));
+
+    long expirationTimeMilliSeconds;
+    if (tokenResponse.getExpiresInSeconds() != null) {
+      expirationTimeMilliSeconds =
+          System.currentTimeMillis() + (tokenResponse.getExpiresInSeconds() * 1000L);
+    } else {
+      logger.atWarning().log(
+          "The OAuth2 provider has returned an access token without a defined expiration (ie `expires_in` was null). "
+              + "We will consider the access token as expired.");
+      expirationTimeMilliSeconds = 0L;
+    }
+    return new AccessToken(tokenResponse.getAccessToken(), expirationTimeMilliSeconds);
+  }
+
+  @Override
+  public Configuration getConf() {
+    return config;
+  }
+
+  @Override
+  public void setConf(Configuration config) {
+    this.config = config;
+  }
+
+  private HttpTransport getTransport() throws IOException {
+    if (httpTransport == null) {
```

Comment:
This seems to be not synchronized field that could be used in concurrent environment, I think it's better to initialize it in constructor and fast fail, which is better too.

Reply:
The issue was with the config, not been loaded to the constructor but through a `setConfig`.